### PR TITLE
JAMES-2520 Deployment test for guice/cassandra/elasticseach docker image

### PR DESCRIPTION
**Pull request by mbaechler**
It's not exactly finished but it is reviewable at least

### CODE REVIEW

File: server/container/testing/src/test/java/org/apache/james/deployment/CassandraContainer.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package org.apache.james.deployment;
```
Comment (chibenwa) [inline server/container/testing/src/test/java/org/apache/james/deployment/CassandraContainer.java]: Licence should come before the package

File: server/container/testing/src/test/java/org/apache/james/deployment/CassandraContainer.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+public class CassandraContainer extends GenericContainer<CassandraContainer>{
+
+    private static final int CASSANDRA_PORT = 9042;
+
+    static final boolean deleteOnExit = false;
+    static final int cassandraMemory = 1000;
+    static final long cassandraContainerMemory = Float.valueOf(cassandraMemory * 1.2f * 1024 * 1024L).longValue();
+    
+    public CassandraContainer() {
+        super("cassandra:2.2.12");
```
Comment (chibenwa) [inline server/container/testing/src/test/java/org/apache/james/deployment/CassandraContainer.java]: now 3.11.3 please

File: server/container/testing/src/test/java/org/apache/james/deployment/CassandraWaitStrategy.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
+package org.apache.james.deployment;
```
Comment (chibenwa) [inline server/container/testing/src/test/java/org/apache/james/deployment/CassandraWaitStrategy.java]: idem

File: server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
+package org.apache.james.deployment;
```
Comment (chibenwa) [inline server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java]: license

File: server/container/testing/src/test/java/org/apache/james/deployment/CassandraWaitStrategy.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import org.rnorth.ducttape.ratelimits.RateLimiter;
+import org.rnorth.ducttape.ratelimits.RateLimiterBuilder;
+import org.rnorth.ducttape.unreliables.Unreliables;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.WaitStrategy;
+import org.testcontainers.containers.wait.strategy.WaitStrategyTarget;
+
+import com.google.common.primitives.Ints;
+
+public class CassandraWaitStrategy implements WaitStrategy {
```
Comment (chibenwa) [inline server/container/testing/src/test/java/org/apache/james/deployment/CassandraWaitStrategy.java]: Isn't it a duplicate from `backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/CassandraWaitStrategy.java`

File: server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+    public GenericContainer<?> cassandra = new CassandraContainer()
+        .withNetwork(network)
+        .withNetworkAliases("cassandra");
+    
+    public GenericContainer<?> elasticsearch = new GenericContainer<>("elasticsearch:2.4.6")
+        .withNetwork(network)
+        .withNetworkAliases("elasticsearch");
+
+    public GenericContainer<?> james = new GenericContainer<>("linagora/james-project:latest")
```
Comment (chibenwa) [inline server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java]: We bqse ourselves on not yet published images... Can't we rather rely on env variables?

File: server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        .withExposedPorts(JMAP_PORT, SMTP_PORT, IMAP_PORT, WEBADMIN_PORT)
+        .withCopyFileToContainer(MountableFile.forClasspathResource("/webadmin.properties"), "/root/conf/")
+        .withCopyFileToContainer(MountableFile.forClasspathResource("/jmap.properties"), "/root/conf/")
+        .withCopyFileToContainer(MountableFile.forClasspathResource("/keystore"), "/root/conf/")
+        .withCopyFileToContainer(MountableFile.forClasspathResource("/jwt_publickey"), "/root/conf/")
+        .withNetwork(network)
+        .withLogConsumer(log -> logger.info(log.getUtf8String()))
+        .waitingFor(new HttpWaitStrategy().forPort(WEBADMIN_PORT).forStatusCodeMatching(status -> true));
```
Comment (chibenwa) [inline server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java]: How do you know webadmin is the last started thing?

IMO we would need a probe that gets its data updated when James is started, and webadmin seems a good idea to expose that probe

File: server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    
+    @Before
+    public void setup() {
+        RestAssured.requestSpecification = new RequestSpecBuilder()
+            .setContentType(ContentType.JSON)
+            .setAccept(ContentType.JSON)
+            .setConfig(newConfig().encoderConfig(encoderConfig().defaultContentCharset(StandardCharsets.UTF_8)))
+            .setPort(james.getMappedPort(WEBADMIN_PORT))
+            .build();
+        jmapApi = new URIBuilder().setScheme("http").setHost(LOCALHOST).setPort(james.getMappedPort(JMAP_PORT));
```
Comment (chibenwa) [inline server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java]: why not use restassured for JMAP?

File: server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+            smtp.connect(LOCALHOST, james.getMappedPort(SMTP_PORT))
+                .authenticate(BART, BART_PASSWORD)
+                .sendMessage(BART, HOMER);
+        }
+    }
+
+    private void assertImapMessageReceived() throws IOException {
+        try (IMAPMessageReader imapMessageReader = new IMAPMessageReader()) {
+            imapMessageReader.connect(LOCALHOST, james.getMappedPort(IMAP_PORT))
+            .login(HOMER, HOMER_PASSWORD)
```
Comment (chibenwa) [inline server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java]: ident

**Comment by chibenwa** [comment]
Cool... But we are testing only **cassandra-guice**. How about other prioducts?

**Comment by chibenwa** [comment]
Don't we only, need to perform this -expensive- test only when validating images we built?

Because playing it while building (ie: we don't have the latest docker images yet) seems a waste of time and pointless...

File: server/container/testing/pom.xml [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+            <artifactId>logback-classic</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.rest-assured</groupId>
+            <artifactId>rest-assured</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
```
Comment (chibenwa) [inline server/container/testing/pom.xml]: We don't use jupiter for new tests?

File: server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java (hunk 7 above)
Comment (mbaechler) [reply server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java]: we should start implementing a health-check endpoint

File: server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java (hunk 6 above)
Comment (mbaechler) [reply server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java]: yes, of course

File: server/container/testing/src/test/java/org/apache/james/deployment/CassandraWaitStrategy.java (hunk 5 above)
Comment (mbaechler) [reply server/container/testing/src/test/java/org/apache/james/deployment/CassandraWaitStrategy.java]: it is

**Comment by mbaechler** [comment]
> Don't we only need to perform this -expensive- test only when validating images we built?

> Because playing it while building (ie: we don't have the latest docker images yet) seems a waste of time and pointless...

Yes, it's supposed to be a specific phase of our build, like today

File: server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java (hunk 7 above)
Comment (chibenwa) [reply server/container/testing/src/test/java/org/apache/james/deployment/DeploymentTest.java]: That's what I started doing...

**Comment by chibenwa** [comment]
I added a profile for it.

File: server/container/testing/pom.xml [hunk 11]
```diff
@@ -33,6 +33,12 @@
     <name>Apache James :: Server :: Deployment Testing</name>
 
     <dependencies>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>apache-james-backends-cassandra</artifactId>
```
Comment (mbaechler) [inline server/container/testing/pom.xml]: this won't work with eclipse : a non-test-jar can't depend on a test-jar anymore

File: server/container/guice/guice-common/src/main/java/org/apache/james/GuiceJamesServer.java [hunk 12]
```diff
@@ -43,31 +44,37 @@
...
 
-    protected GuiceJamesServer(Module module) {
+    protected GuiceJamesServer(IsStartedProbe isStartedProbe, Module module) {
+        this.isStartedProbe = isStartedProbe;
         this.module = module;
     }
     
     public GuiceJamesServer combineWith(Module... modules) {
-        return new GuiceJamesServer(Modules.combine(Iterables.concat(Arrays.asList(module), Arrays.asList(modules))));
+        return new GuiceJamesServer(isStartedProbe, Modules.combine(Iterables.concat(Arrays.asList(module), Arrays.asList(modules))));
```
Comment (mbaechler) [inline server/container/guice/guice-common/src/main/java/org/apache/james/GuiceJamesServer.java]: you should not have to repeat yourself for that

File: server/container/guice/guice-common/src/main/java/org/apache/james/IsStartedProbe.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james;
+
+import org.apache.james.utils.GuiceProbe;
+
+public class IsStartedProbe implements GuiceProbe {
+
+    private boolean isStarted;
```
Comment (mbaechler) [inline server/container/guice/guice-common/src/main/java/org/apache/james/IsStartedProbe.java]: should be `volatile`

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/CassandraJamesServerMain.java [hunk 14]
```diff
@@ -71,7 +72,8 @@
         new SwaggerRoutesModule(),
         new WebAdminServerModule(),
         new DLPRoutesModule(),
-        new SieveQuotaRoutesModule());
+        new SieveQuotaRoutesModule(),
+        new StatusRouteModule());
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/CassandraJamesServerMain.java]: the idea is really great but we should generalize that to a real healthcheck endpoint IMO

File: server/container/testing/pom.xml (hunk 11 above)
Comment (chibenwa) [reply server/container/testing/pom.xml]: Depedency removed in the end...

**Comment by chibenwa** [comment]
Most of my work is exposed here: https://github.com/linagora/james-project/pull/1646

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/CassandraJamesServerMain.java (hunk 14 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/CassandraJamesServerMain.java]: What do you mean? `red/yellow/green` with various components defining their health and James doing a logical AND?

**Comment by mbaechler** [comment]
> I added a profile for it.

Oh no, not the return of the profiles !

Please use a junit tag instead

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/CassandraJamesServerMain.java (hunk 14 above)
Comment (mbaechler) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/CassandraJamesServerMain.java]: There are plenty of way of doing that. Spring boot has a lot of built-in healthcheck features.
dropwizard has some too ( https://www.dropwizard.io/1.3.5/docs/manual/core.html#health-checks )
OBM has some : https://www.obm.org/wiki/obm-health-check-system
If you look at k8s documentation, any endpoint returning less than HTTP 400 on success is a good enough liveness probe.

IMO, having a simple health probe registry and a specific endpoint (no admin access required but maybe a specific port) to run the probes would be enough.

**Comment by aduprat** [comment]
https://github.com/linagora/james-project/pull/2061

**Pull request closed without merging**